app/code/Magento/InventoryCatalog/Model/LegacySynchronization/ToLegacyCatalogInventory/SetDataToLegacyInventory.php

```diff
@@ -0,0 +1,131 @@
+use Magento\InventoryApi\Api\Data\SourceItemInterface;
+use Magento\InventoryCatalog\Model\GetProductIdsBySkus;
+use Magento\InventoryCatalog\Model\LegacySynchronization\GetLegacyStockItemsByProductIds;
+use Magento\InventoryCatalog\Model\ResourceModel\SetDataToLegacyStockItem;
+use Magento\InventoryCatalogApi\Api\DefaultSourceProviderInterface;
+
+class SetDataToLegacyInventory
+{
+    /**
+     * @var SetDataToLegacyStockItem
+     */
+    private $setDataToLegacyStockItem;
+
+    /**
+     * @var GetLegacyStockItemsByProductIds
+     */
+    private $getLegacyStockItemsByProductIds;
+
+    /**
+     * @var StockStateProviderInterface
+     */
+    private $stockStateProvider;
+
+    /**
+     * @var Processor
+     */
+    private $indexerProcessor;
+
+    /**
+     * @var GetProductIdsBySkus
+     */
+    private $getProductIdsBySkus;
+
+    /**
+     * @var DefaultSourceProviderInterface
+     */
+    private $defaultSourceProvider;
+
+    /**
+     * SetDataToLegacyCatalogInventory constructor.
+     * @param DefaultSourceProviderInterface $defaultSourceProvider
+     * @param SetDataToLegacyStockItem $setDataToLegacyStockItem
+     * @param GetLegacyStockItemsByProductIds $getLegacyStockItemsByProductIds
+     * @param StockStateProviderInterface $stockStateProvider
+     * @param Processor $indexerProcessor
+     * @param GetProductIdsBySkus $getProductIdsBySkus
+     * @SuppressWarnings(PHPMD.LongVariable)
+     */
+    public function __construct(
+        DefaultSourceProviderInterface $defaultSourceProvider,
+        SetDataToLegacyStockItem $setDataToLegacyStockItem,
+        GetLegacyStockItemsByProductIds $getLegacyStockItemsByProductIds,
+        StockStateProviderInterface $stockStateProvider,
+        Processor $indexerProcessor,
+        GetProductIdsBySkus $getProductIdsBySkus
+    ) {
+        $this->setDataToLegacyStockItem = $setDataToLegacyStockItem;
+        $this->getLegacyStockItemsByProductIds = $getLegacyStockItemsByProductIds;
+        $this->stockStateProvider = $stockStateProvider;
+        $this->indexerProcessor = $indexerProcessor;
+        $this->getProductIdsBySkus = $getProductIdsBySkus;
+        $this->defaultSourceProvider = $defaultSourceProvider;
+    }
+
+    /**
+     * Synchronously execute legacy alignment
+     *
+     * @param array $sourceItemsData
+     * @throws LocalizedException
+     */
+    public function execute(array $sourceItemsData): void
+    {
+        $productIds = [];
+        foreach ($sourceItemsData as $sourceItemData) {
+            $sku = (string) $sourceItemData[SourceItemInterface::SKU];
+
+            if ($sourceItemData[SourceItemInterface::SOURCE_CODE] !== $this->defaultSourceProvider->getCode()) {
+                throw new LocalizedException(__('Only default source can synchronize legacy stock'));
+            }
+
+            try {
+                $productId = (int) current($this->getProductIdsBySkus->execute([$sku]));
```

Comment:
it's better not to call `$this->getProductIdsBySkus->execute([$sku])` for each SKU, but collect all SKUs and proceed with one call instead.

As potentially each call would access DB in a separate MySQL query

Reply:
Maybe using `\Magento\Catalog\Model\ResourceModel\Product`. Single SKu processing was in place to avoid inconsistency issues with source items for non existing products.